GVFS/GVFS.Common/NuGetUpgrader/NuGetFeed.cs

```diff
@@ -0,0 +1,224 @@
+        /// <returns>List of packages that match query parameters</returns>
+        public virtual async Task<IList<IPackageSearchMetadata>> QueryFeedAsync(string packageId)
+        {
+            try
+            {
+                PackageMetadataResource packageMetadataResource = await this.sourceRepository.GetResourceAsync<PackageMetadataResource>();
+                IEnumerable<IPackageSearchMetadata> queryResults = await packageMetadataResource.GetMetadataAsync(
+                    packageId,
+                    includePrerelease: true,
+                    includeUnlisted: true,
+                    sourceCacheContext: this.sourceCacheContext,
+                    log: this.nuGetLogger,
+                    token: CancellationToken.None);
+
+                return queryResults.ToList();
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"Failed to query the NuGet package feed due to error: {ex.Message}", ex);
+            }
+        }
+
+        /// <summary>
+        /// Download the specified packageId from the NuGet feed.
+        /// </summary>
+        /// <param name="packageId">PackageIdentity to download.</param>
+        /// <returns>Path to the downloaded package.</returns>
+        public virtual async Task<string> DownloadPackage(PackageIdentity packageId)
+        {
+            string downloadPath = Path.Combine(this.downloadFolder, $"{this.feedName}.zip");
+            PackageDownloadContext packageDownloadContext = new PackageDownloadContext(
+                this.sourceCacheContext,
+                this.downloadFolder,
+                true);
+
+            DownloadResource downloadResource = await this.sourceRepository.GetResourceAsync<DownloadResource>();
+
+            using (DownloadResourceResult downloadResourceResult = await downloadResource.GetDownloadResourceResultAsync(
+                       packageId,
+                       packageDownloadContext,
+                       globalPackagesFolder: string.Empty,
+                       logger : this.nuGetLogger,
+                       token: CancellationToken.None))
+            {
+                // Check download result status
+                using (FileStream fileStream = File.Create(downloadPath))
+                {
+                    downloadResourceResult.PackageStream.CopyTo(fileStream);
+                }
+            }
+
+            return downloadPath;
+        }
+
+        /// <summary>
+        /// Implementation of logger used by NuGet library. It takes all output
+        /// and redirects it to the GVFS logger.
+        /// </summary>
+        public class Logger : ILogger
+        {
+            private ITracer tracer;
+
+            public Logger(ITracer tracer)
+            {
+                this.tracer = tracer;
+            }
+
+            public void Log(LogLevel level, string data)
+            {
+                string message = $"NuGet Logger: ({level}): {data}";
```

Comment:
Should the `data` be put in an `EventMetadata` field?  I don't think it matters too much.